{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#This script uses primer3 to calculate free-energy of dimerization for all the primers generated in the previous script. \n",
    "#This script requires a local installation of primer3.\n",
    "\n",
    "import os\n",
    "import subprocess\n",
    "from time import time\n",
    "\n",
    "#Defining classes to help with storing primer information\n",
    "class gene:\n",
    "    def __init__(self, name, primers):\n",
    "        self.name = name\n",
    "        self.primers = primers\n",
    "        return\n",
    "    \n",
    "class primer_pair:\n",
    "    def __init__(self, A, B, name, count):\n",
    "        self.id = f\"{name}-{count}\"\n",
    "        self.gene = name\n",
    "        self.fseq = A.split(\"+\")[0]\n",
    "        self.rseq = A.split(\"+\")[3]\n",
    "        self.fstart = A.split(\"+\")[2]\n",
    "        self.penalty = B.split(\"-\")[0]\n",
    "        self.ftm = B.split(\"-\")[1]\n",
    "        self.rtm = B.split(\"-\")[2]\n",
    "        self.ampliconlen = B.split(\"-\")[3]\n",
    "        self.path = []\n",
    "        self.neighbours = []\n",
    "        return\n",
    "            \n",
    "\n",
    "#specify the location of the output of the primer generator script\n",
    "os.chdir(\"/home/flan3@ad.wisc.edu/MplexPrimerPicker/projects/mockcommunity/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sepiderm-dnaG\n",
      "Pputida-dnaG\n",
      "Llactis-dnaG\n",
      "Bfragilis-rpoB\n",
      "Chiranonis-rpoB\n",
      "Ecoli-rpoB\n",
      "Bsubtilis-rpoB\n",
      "RFP\n",
      "P5/P7\n",
      "16S\n",
      "802 total primers\n"
     ]
    }
   ],
   "source": [
    "#load the primer lists generated by the last script\n",
    "import pickle\n",
    "with open(\"Primers.pickle\", \"rb\") as infile:\n",
    "    genelist = pickle.load(infile)\n",
    "\n",
    "#format the list of genes and fix primer names\n",
    "glist = []\n",
    "count = 0\n",
    "for g in genelist:\n",
    "    plist = []\n",
    "    for p in genelist[g]:\n",
    "        plist.append(p)\n",
    "        count+= 1\n",
    "    glist.append(gene(p.gene.split(\"|\")[-2:], plist))\n",
    "    print(p.gene)\n",
    "genelist = glist\n",
    "print(count, \"total primers\")\n",
    "\n",
    "#truncate primers from 3' end to 60bp because ntthal the thermodynamics calculator cannot take >60bp long oligos.\n",
    "for g in genelist:\n",
    "    for p in g.primers:\n",
    "        if len(p.fseq) > 60: #ntthal can't take longer than 60bp long\n",
    "            p.fseq = p.fseq[len(p.fseq)-60:] #so we truncate from 5'            \n",
    "        if len(p.rseq) > 60: #ntthal can't take longer than 60bp long\n",
    "\n",
    "            p.rseq = p.rseq[len(p.rseq)-60:] #so we truncate from 5'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#define functions to run ntthal (the program that checks primer dimer deltaGs)\n",
    "\n",
    "#command to run ntthal, the thermodynamic calculator of primer3\n",
    "#ntthal is located in the primer3 installation directory as shown below\n",
    "ntthal = \"~/primer3/src/ntthal -mv 200 -dv 3 -n 0.2 -t 25 -d 50\"  #calculate at 25C, 200mM Naccl, 3mM Mgcl, 0.2mM dNTP, 60oC temp, 50nM primers\n",
    "\n",
    "#This function runs ntthal for 2 input sequences, comparing the deltaG of hybridization of end-hybridization of both primers to another\n",
    "#This function returns the lowest deltaG from the two comparisons\n",
    "def run_ntthal(s1, s2):\n",
    "    try:\n",
    "        stream = os.popen(f\"{ntthal} -s1 {s1} -s2 {s2} -a END1\")\n",
    "        l = stream.readlines()[0].split()\n",
    "        score1 = float(l[13])\n",
    "    except:\n",
    "        print(stream.readlines())\n",
    "        score1 = 999\n",
    "    \n",
    "    try:\n",
    "        stream = os.popen(f\"{ntthal} -s1 {s1} -s2 {s2} -a END2\")\n",
    "        l = stream.readlines()[0].split()\n",
    "        score2 = float(l[13])\n",
    "    except:\n",
    "        print(stream.readlines())\n",
    "        score2 = 999    \n",
    "        \n",
    "    return min(score1, score2)\n",
    "\n",
    "#function to check the dimer deltaGs between ith gene in the list against other genes in the list (that's 4 primers cross checked)\n",
    "#stores the results in an adjacency dictionary of (primerset1,primerset2):mindeltaG\n",
    "def check_genes(gene, i):\n",
    "    adjdict = {} #create a local dict to store these things\n",
    "    for p in gene.primers: #loop through each primer under gene i\n",
    "        seqs = [0,0,0,0]\n",
    "        ID = p.id\n",
    "        seqs[0] = p.fseq\n",
    "        seqs[1] = p.rseq\n",
    "        homoGs = []\n",
    "\n",
    "        #first check self-dimerization once so we don't keep repeating it\n",
    "        for s in seqs[0:2]:\n",
    "            homoGs.append(run_ntthal(s, s))\n",
    "            \n",
    "        for j in range(i+1, len(genelist)): #for each other gene\n",
    "            for q in genelist[j].primers: #loop through primers under other genes\n",
    "                #ts = process_time()\n",
    "                ID2 = q.id\n",
    "                seqs[2] = q.fseq\n",
    "                seqs[3] = q.rseq          \n",
    "                deltaGs = list(homoGs)\n",
    "\n",
    "                #do all-way comparisons (completely check for dimers self and hetero)\n",
    "                #check homodimerization for the other 2 sequences\n",
    "                for s in seqs[2:4]:\n",
    "                    deltaGs.append(run_ntthal(s, s))\n",
    "                #now check for cross-dimerization\n",
    "                for z in range(4):\n",
    "                    for y in range(4):\n",
    "                        if z != y: #skip the homo check steps\n",
    "                            deltaGs.append(run_ntthal(seqs[y], seqs[z])) #check all heterocombinations, AB != BA when doing End comparisons\n",
    "                \n",
    "                #add to the adjacency dictionary\n",
    "                adjdict[frozenset((ID,ID2))] = min(deltaGs) #add the worst deltaGs\n",
    "                #print(process_time()-ts) - timing one primer-v-primer comparison - takes 0.015s\n",
    "    return adjdict\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#parallelized calculation of thermodynamic binding energies for primer pairs using the above functions\n",
    "import multiprocessing\n",
    "from joblib import Parallel, delayed\n",
    "\n",
    "num_cores = multiprocessing.cpu_count() #number of cores to use, defaults to all cores\n",
    "\n",
    "t1 = time()  #to keep track of how long the job took\n",
    "adjdictlist = Parallel(n_jobs=10)(delayed(check_genes)(genelist[i], i) for i in range(len(genelist))) #do the parallelized adjmatrix calculations\n",
    "\n",
    "#merge all outputs of different jobs into a single datastructure\n",
    "adjdict = {}\n",
    "for d in adjdictlist:\n",
    "    adjdict.update(d)\n",
    "\n",
    "#store results on harddrive as a pickle, this file will be used by the third script to pick primer pools\n",
    "import pickle\n",
    "with open(\"dimerization-deltaGs.pickle\", \"wb\") as outfile:\n",
    "    pickle.dump(adjdict, outfile)\n",
    "\n",
    "t2 = time() #keep track of how long the job took\n",
    "    \n",
    "print(f\"running on {num_cores} took {t2 - t1} seconds\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#visualize the distribution of deltaGs \n",
    "import matplotlib.pyplot as plot\n",
    "\n",
    "#check the distributions of deltaGs and Tms\n",
    "dGs =[]\n",
    "\n",
    "for v in adjdict.values():\n",
    "    dGs.append(v)\n",
    "\n",
    "fig, axs = plot.subplots()\n",
    "fig.tight_layout()\n",
    "\n",
    "axs.hist(dGs, bins = 50)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
